server.go

```diff
@@ -242,15 +211,94 @@ func filterLayersByZoom(inLayers []Layer, z int) []Layer {
 	return outLayers
 }
 
+// getLayerDataFromSource retrieves layer data from the original backend source
+func (s *Server) getLayerDataFromSource(ctx context.Context, layer Layer, req *TileRequest) (*mvt.Layer, error) {
+	fc, err := layer.GetFeatures(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	fcLayer := mvt.NewLayer(layer.Name, fc)
+	fcLayer.Version = 2 // Set to tile spec v2
+	fcLayer.ProjectToTile(req.MapTile())
+	fcLayer.Clip(mvt.MapboxGLDefaultExtentBound)
+	return fcLayer, nil
+}
+
+// getLayerDataFromCache retrieves layer data from the configured cache
+func (s *Server) getLayerDataFromCache(ctx context.Context, cacheKey string) (*mvt.Layer, error) {
+	raw, err := s.Cache.Get(cacheKey)
+	if err != nil {
+		return nil, err
+	}
+	layers, err := mvt.UnmarshalGzipped(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note: since we store an mvt.Layers array with only a single layer, we need to pull out
+	// the first layer to match the return interface
+	return layers[0], nil
+}
+
+// getLayerData retrieves layer data either from cache or the original source
+func (s *Server) getLayerData(ctx context.Context, layer Layer, req *TileRequest) (*mvt.Layer, error) {
+	cacheKey := fmt.Sprintf("%s/%s", layer.String(), req.String())
+	if layer.Cacheable && s.Cache.Exists(cacheKey) {
+		Logger.Debugf("Key [%s] found in cache", cacheKey)
+		if fcLayer, err := s.getLayerDataFromCache(ctx, cacheKey); err == nil {
+			return fcLayer, nil
+		} else {
+			Logger.Warningf("Failed to retrieve layer data from cache [%s]: %s", cacheKey, err)
```

Comment:
When would this behavior occur?

Reply:
A handful of failure scenarios are included in the PR description above:

> in this revised implementation, cache retrieval failures are treated the same as a cache miss, which should fix a broad class of potential issues (e.g. temporary network failures, bad data/deserialization issues, etc.)

To highlight a couple:

* One concrete scenario that would fail today but would work with this implementation is if we for whatever reason lost our Redis cluster (e.g. Kubernetes decideds to restart it, or we decide to reconfigure it live, etc.); this is a common enough scenario that we should be resilient to.
* Another scenario is more related to future-proofing: if we decided to serialize the cached layer data in Redis in a different way (e.g. using `gob` instead of GZIPed MVTs), then any older cache data may fail to deserialize from Redis; this change allows a bit more flexibility in that scenario by simply warning about the cache deserialization failure and re-pulling the data from the backend source, then implicitly "upgrading" the cache data by storing in the more current deserialization format